### PR TITLE
Add typescript support

### PR DESCRIPTION
**Pull request by GregTheGreek**
## Description

This PR focuses on adding TypeScript in an incremental way. Currently `allowJs` is enabled allowing us to write TypeScript and JavaScript side by side. The configuration is very lenient, and in the future as we make progress on the typescript interfaces we should tighten it up.

It also:
- Changes the build to build using tsc (future change to babel?)
- Changes the tests to run the tsnode/register so typescript is supported
-  Adds support for up-to es8 (classes, arrow functions, etc...)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ x ] I have selected the correct base branch.
- [ x ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ x ] My changes generate no new warnings.
- [ x ] Any dependent changes have been merged and published in downstream modules.
- [ x ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ x ] I ran `npm run test:unit` with success.
- [ x ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ x ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [ x ] I have tested my code on the live network.
- [ x ] I have checked the Deploy Preview and it looks correct.
- [ x ] I have updated the `CHANGELOG.md` file in the root folder.

### CODE REVIEW

File: packages/web3-core/src/index.js [hunk 1]
```diff
@@ -20,67 +20,70 @@
  * @date 2017
  */
 
-"use strict";
+// import requestManager from "web3-core-requestmanager";
```
Comment (Spacesai1or) [inline packages/web3-core/src/index.js]: Remove if not needed (line below as well)

**Review by Spacesai1or — DISMISSED** [review]
Other than the two commented imports, lgtm!

File: scripts/e2e.ganache.sh [hunk 2]
```diff
@@ -26,7 +26,8 @@ echo " "
 npx ganache-cli --noVMErrorsOnRPCResponse > /dev/null &
 client=$!
 npx wait-port 8545
-
+ls -la /home/runner/work/web3.js/web3.js/packages/web3/
+ls -la /home/runner/work/web3.js/web3.js/packages/web3/lib
```
Comment (ryanio) [inline scripts/e2e.ganache.sh]: should this still be here? i assume it was for some debugging

**Review by Spacesai1or — DISMISSED** [review]

File: scripts/e2e.ganache.sh (hunk 2 above)
Comment (GregTheGreek) [reply scripts/e2e.ganache.sh]: :facepalm: woops

**Comment by GregTheGreek** [comment]
@spacesailor24 and @ryanio I updated the changelog and addressed a previous comment. Ready for re-review.

**Comment by ryanio** [comment]
@GregTheGreek looks good 👍 in the past we have only merged changes to `dist/web3.min.js` on release (like noted [here](https://github.com/ethereum/web3.js/pull/3618#pullrequestreview-443378538)), what do you think?

**Comment by GregTheGreek** [comment]
> @GregTheGreek looks good  in the past we have only merged changes to `dist/web3.min.js` on release (like noted [here](https://github.com/ethereum/web3.js/pull/3618#pullrequestreview-443378538)), what do you think?

hmm let me look at that, i thought i pulled in the ones from `1.x`

**Review by ryanio — APPROVED** [review]
LGTM!

**Pull request merged**